README.md
# Weather

[![NPM][npm-image]][npm-url] [![Build Status][travis-image]][travis-url]

Weather is a module for obtaining weather information.

## Installation

```bash
npm install weather-js
```

## Usage

```javascript
var weather = require('weather-js');

// Options:
// search:     location name or zipcode
// degreeType: F or C

weather.find({search: 'San Francisco, CA', degreeType: 'F'}, function(err, result) {
  if(err) console.log(err);

  console.log(JSON.stringify(result, null, 2));
});
/*
[
  {
    "location": {
      "name": "San Francisco, CA",
      "lat": "37.777",
      "long": "-122.42",
      "timezone": "-8",
      "alert": "",
      "degreetype": "F",
      "imagerelativeurl": "http://blob.weather.microsoft.com/static/weather4/en-us/"
    },
    "current": {
      "temperature": "50",
      "skycode": "30",
      "skytext": "Partly Sunny",
      "date": "2017-03-05",
      "observationtime": "16:35:00",
      "observationpoint": "San Francisco, CA",
      "feelslike": "45",
      "humidity": "58",
      "winddisplay": "13 mph West",
      "day": "Sunday",
      "shortday": "Sun",
      "windspeed": "13 mph",
      "imageUrl": "http://blob.weather.microsoft.com/static/weather4/en-us/law/30.gif"
    },
    "forecast": [
      {
        "low": "47",
        "high": "59",
        "skycodeday": "11",
        "skytextday": "Rain",
        "date": "2017-03-04",
        "day": "Saturday",
        "shortday": "Sat",
        "precip": ""
      },
      {
        "low": "46",
        "high": "52",
        "skycodeday": "34",
        "skytextday": "Mostly Sunny",
        "date": "2017-03-05",
        "day": "Sunday",
        "shortday": "Sun",
        "precip": "40"
      },
      {
        "low": "47",
        "high": "56",
        "skycodeday": "28",
        "skytextday": "Mostly Cloudy",
        "date": "2017-03-06",
        "day": "Monday",
        "shortday": "Mon",
        "precip": "60"
      },
      {
        "low": "49",
        "high": "58",
        "skycodeday": "28",
        "skytextday": "Mostly Cloudy",
        "date": "2017-03-07",
        "day": "Tuesday",
        "shortday": "Tue",
        "precip": "10"
      },
      {
        "low": "52",
        "high": "62",
        "skycodeday": "32",
        "skytextday": "Sunny",
        "date": "2017-03-08",
        "day": "Wednesday",
        "shortday": "Wed",
        "precip": "10"
      }
    ]
  }
]
*/
```

## Notes

- It uses `weather.service.msn.com`

## License

Licensed under The MIT License (MIT)  
For the full copyright and license information, please view the LICENSE.txt file.

[npm-url]: http://npmjs.org/package/weather-js
[npm-image]: https://badge.fury.io/js/weather-js.svg

[travis-url]: https://travis-ci.org/devfacet/weather
[travis-image]: https://travis-ci.org/devfacet/weather.svg?branch=master


CHANGELOG.md
# Changelog

## 1.0.3 (2017-03-05)

- Maintenance

## 1.0.2 (2015-03-15)

- Misc. updates

## 1.0.1 (2014-12-12)

- Change request url

## 1.0.0 (2014-07-04)

- Stable release

## 0.1.1 (2014-06-05)

- Add image url

## 0.1.0 (2014-06-04)

- Initial release

test/test-all.js
/* jslint node: true, sub: true */
/* global describe: false, it: false */
'use strict';

var weather = require('../'),
    expect  = require('chai').expect;

// Tests

describe('weather', function() {

  // find
  describe('find()', function() {
    it('should find a location with weather information', function(done) {
      weather.find({search: 'San Francisco, CA', degreeType: 'F'}, function(err, result) {
        if(err) return done(err);

        expect(err).to.be.equal(undefined);

        expect(result).to.be.a('array');
        expect(result).to.have.property('length').to.be.equal(1);
        expect(result[0]).to.be.a('object');

        expect(result[0]).to.have.property('location').to.be.a('object');
        expect(result[0]['location']).to.have.property('name', 'San Francisco, CA');
        expect(result[0]['location']).to.have.property('lat', '37.777');
        expect(result[0]['location']).to.have.property('long', '-122.42');
        expect(result[0]['location']).to.have.property('timezone').to.be.a('string');
        expect(result[0]['location']).to.have.property('alert').to.be.a('string');
        expect(result[0]['location']).to.have.property('degreetype', 'F');
        expect(result[0]['location']).to.have.property('imagerelativeurl').to.be.a('string');

        expect(result[0]).to.have.property('current').to.be.a('object');
        expect(result[0]['current']).to.have.property('temperature');
        expect(result[0]['current']).to.have.property('temperature');
        expect(result[0]['current']).to.have.property('temperature');
        expect(result[0]['current']).to.have.property('skycode');
        expect(result[0]['current']).to.have.property('skytext');
        expect(result[0]['current']).to.have.property('date');
        expect(result[0]['current']).to.have.property('observationtime');
        expect(result[0]['current']).to.have.property('observationpoint');
        expect(result[0]['current']).to.have.property('feelslike');
        expect(result[0]['current']).to.have.property('humidity');
        expect(result[0]['current']).to.have.property('winddisplay');
        expect(result[0]['current']).to.have.property('day');
        expect(result[0]['current']).to.have.property('shortday');
        expect(result[0]['current']).to.have.property('windspeed');
        expect(result[0]['current']).to.have.property('imageUrl');

        expect(result[0]).to.have.property('forecast').to.be.a('array');
        expect(result[0]['forecast']).to.have.property('length').to.be.above(0);
        expect(result[0]['forecast'][0]).to.be.a('object');
        expect(result[0]['forecast'][0]).to.have.property('low');
        expect(result[0]['forecast'][0]).to.have.property('high');
        expect(result[0]['forecast'][0]).to.have.property('skycodeday');
        expect(result[0]['forecast'][0]).to.have.property('skytextday');
        expect(result[0]['forecast'][0]).to.have.property('date');
        expect(result[0]['forecast'][0]).to.have.property('day');
        expect(result[0]['forecast'][0]).to.have.property('shortday');
        expect(result[0]['forecast'][0]).to.have.property('precip');

        done();
      });
    });

    it('should find multiple locations with weather information', function(done) {
      weather.find({search: 'San Francisco', degreeType: 'F'}, function(err, result) {
        if(err) return done(err);

        expect(err).to.be.equal(undefined);
        expect(result).to.be.a('array');
        expect(result).to.have.property('length').to.be.above(1);
        done();
      });
    });

    it('should fail to find a location (missing options)', function(done) {
      weather.find(null, function(err, result) {
        if(!err) return done('No error!');

        expect(result).to.be.equal(undefined);
        done();
      });
    });
  });

});

index.js
/*
 * Weather
 * For the full copyright and license information, please view the LICENSE.txt file.
 */

/* jslint node: true, sub: true */
'use strict';

var request = require('request'),
    qs      = require('querystring'),
    xml2JS  = require('xml2js');

// Init the module
module.exports = (function() {

  var xmlParser     = new xml2JS.Parser({charkey: 'C$', attrkey: 'A$', explicitArray: true}),
      defLang       = 'en-US',
      defDegreeType = 'F',
      defTimeout    = 10000,
      findUrl       = 'http://weather.service.msn.com/find.aspx',
      find;         // finds and gets weather information - function

  find = function find(options, callback) {

    if(typeof callback !== 'function')
      callback = function callback(err, result) { return err || result; };

    if(!options || typeof options !== 'object')
      return callback('Invalid options!');

    if(!options.search)
      return callback('Missing search input!');

    var result,
        lang       = options.lang || defLang,
        degreeType = options.degreeType || defDegreeType,
        timeout    = options.timeout || defTimeout,
        search     = qs.escape(''+options.search),
        reqUrl     = findUrl + '?src=outlook&weadegreetype=' + (''+degreeType) + '&culture=' + (''+lang) + '&weasearchstr=' + search;

    request.get({url: reqUrl, timeout: timeout}, function(err, res, body) {

      if(err) return callback(err);
      if(res.statusCode !== 200) return callback('Request failed (' + res.statusCode + ')');

      xmlParser.parseString(body, function(err, resultJSON) {
        if(err) return callback(err);

        if(!resultJSON || !resultJSON.weatherdata || !resultJSON.weatherdata.weather)
          return callback('Unexpected error! Invalid content.');

        if(resultJSON.weatherdata.weather['A$'] && resultJSON.weatherdata.weather['A$'].errormessage)
          return callback(resultJSON.weatherdata.weather['A$'].errormessage);

        if(!(resultJSON.weatherdata.weather instanceof Array)) {
          return callback('Unexpected error! Missing weather info.');
        }

        var weatherLen = resultJSON.weatherdata.weather.length,
            weatherItem;
        for(var i = 0; i < weatherLen; i++) {

          if(typeof resultJSON.weatherdata.weather[i]['A$'] !== 'object')
            continue;

          weatherItem = {
            location: {
              name: resultJSON.weatherdata.weather[i]['A$']['weatherlocationname'],
              zipcode: resultJSON.weatherdata.weather[i]['A$']['zipcode'],
              lat: resultJSON.weatherdata.weather[i]['A$']['lat'],
              long: resultJSON.weatherdata.weather[i]['A$']['long'],
              timezone: resultJSON.weatherdata.weather[i]['A$']['timezone'],
              alert: resultJSON.weatherdata.weather[i]['A$']['alert'],
              degreetype: resultJSON.weatherdata.weather[i]['A$']['degreetype'],
              imagerelativeurl: resultJSON.weatherdata.weather[i]['A$']['imagerelativeurl']
              //url: resultJSON.weatherdata.weather[i]['A$']['url'],
              //code: resultJSON.weatherdata.weather[i]['A$']['weatherlocationcode'],
              //entityid: resultJSON.weatherdata.weather[i]['A$']['entityid'],
              //encodedlocationname: resultJSON.weatherdata.weather[i]['A$']['encodedlocationname']
            },
            current: null,
            forecast: null
          };

          if(resultJSON.weatherdata.weather[i]['current'] instanceof Array && resultJSON.weatherdata.weather[i]['current'].length > 0) {
            if(typeof resultJSON.weatherdata.weather[i]['current'][0]['A$'] === 'object') {
              weatherItem.current = resultJSON.weatherdata.weather[i]['current'][0]['A$'];

              weatherItem.current.imageUrl = weatherItem.location.imagerelativeurl + 'law/' + weatherItem.current.skycode + '.gif';
            }
          }

          if(resultJSON.weatherdata.weather[i]['forecast'] instanceof Array) {
            weatherItem.forecast = [];
            for(var k = 0; k < resultJSON.weatherdata.weather[i]['forecast'].length; k++) {
              if(typeof resultJSON.weatherdata.weather[i]['forecast'][k]['A$'] === 'object')
                weatherItem.forecast.push(resultJSON.weatherdata.weather[i]['forecast'][k]['A$']);
            }
          }

          if(!result) result = [];
          result.push(weatherItem);
        }

        return callback(undefined, result);
      });
    });
  };

  // Return
  return {
    find: find
  };
})();

lib/weather.js
/*
 * Weather
 * Copyright (c) 2014 Fatih Cetinkaya (http://github.com/cmfatih/weather)
 * For the full copyright and license information, please view the LICENSE.txt file.
 */

/* jslint node: true */
'use strict';

var request = require('request'),
    qs      = require('querystring'),
    xml2JS  = require('xml2js');

// Init the module
exports = module.exports = (function() {

  var xmlParser     = new xml2JS.Parser({charkey: 'C$', attrkey: 'A$', explicitArray: true}),
      defLang       = 'en-US',
      defDegreeType = 'F',
      defTimeout    = 10000,
      findUrl       = 'http://weather.service.msn.com/find.aspx',
      find;         // finds and gets weather information - function

  find = function find(options, callback) {

    if(typeof callback !== 'function')
      callback = function callback(err, result) { return err || result; };

    if(!options || typeof options !== 'object')
      return callback('Invalid options!');

    if(!options.search)
      return callback('Missing search input!');

    var result,
        lang       = options.lang || defLang,
        degreeType = options.degreeType || defDegreeType,
        timeout    = options.timeout || defTimeout,
        search     = qs.escape(''+options.search),
        reqUrl     = findUrl + '?src=outlook&weadegreetype=' + (''+degreeType) + '&culture=' + (''+lang) + '&weasearchstr=' + search;

    request.get({url: reqUrl, timeout: timeout}, function(err, res, body) {

      if(err) return callback(err);
      if(res.statusCode !== 200) return callback('Request failed (' + res.statusCode + ')');

      xmlParser.parseString(body, function(err, resultJSON) {
        if(err) return callback(err);

        if(!resultJSON.weatherdata || !resultJSON.weatherdata.weather)
          return callback('Unexpected error! Invalid content.');

        if(resultJSON.weatherdata.weather['A$'] && resultJSON.weatherdata.weather['A$'].errormessage)
          return callback(resultJSON.weatherdata.weather['A$'].errormessage);

        if(!(resultJSON.weatherdata.weather instanceof Array)) {
          return callback('Unexpected error! Missing weather info.');
        }

        var weatherLen = resultJSON.weatherdata.weather.length,
            weatherItem;
        for(var i = 0; i < weatherLen; i++) {

          if(typeof resultJSON.weatherdata.weather[i]['A$'] !== 'object')
            continue;

          weatherItem = {
            location: {
              name: resultJSON.weatherdata.weather[i]['A$']['weatherlocationname'],
              zipcode: resultJSON.weatherdata.weather[i]['A$']['zipcode'],
              lat: resultJSON.weatherdata.weather[i]['A$']['lat'],
              long: resultJSON.weatherdata.weather[i]['A$']['long'],
              timezone: resultJSON.weatherdata.weather[i]['A$']['timezone'],
              alert: resultJSON.weatherdata.weather[i]['A$']['alert'],
              degreetype: resultJSON.weatherdata.weather[i]['A$']['degreetype'],
              imagerelativeurl: resultJSON.weatherdata.weather[i]['A$']['imagerelativeurl']
              //url: resultJSON.weatherdata.weather[i]['A$']['url'],
              //code: resultJSON.weatherdata.weather[i]['A$']['weatherlocationcode'],
              //entityid: resultJSON.weatherdata.weather[i]['A$']['entityid'],
              //encodedlocationname: resultJSON.weatherdata.weather[i]['A$']['encodedlocationname']
            },
            current: null,
            forecast: null
          };

          if(resultJSON.weatherdata.weather[i]['current'] instanceof Array && resultJSON.weatherdata.weather[i]['current'].length > 0) {
            if(typeof resultJSON.weatherdata.weather[i]['current'][0]['A$'] === 'object') {
              weatherItem.current = resultJSON.weatherdata.weather[i]['current'][0]['A$'];

              weatherItem.current.imageUrl = weatherItem.location.imagerelativeurl + 'law/' + weatherItem.current.skycode + '.gif';
            }
          }

          if(resultJSON.weatherdata.weather[i]['forecast'] instanceof Array) {
            weatherItem.forecast = [];
            for(var k = 0; k < resultJSON.weatherdata.weather[i]['forecast'].length; k++) {
              if(typeof resultJSON.weatherdata.weather[i]['forecast'][k]['A$'] === 'object')
                weatherItem.forecast.push(resultJSON.weatherdata.weather[i]['forecast'][k]['A$']);
            }
          }

          if(!result) result = [];
          result.push(weatherItem);
        }

        return callback(undefined, result);
      });
    });
  };

  // Return
  return {
    find: find
  };
})();